spec/consensus/light-client/detection.md

```diff
@@ -1,3 +1,459 @@
+
+
+
+#### **[LCD-REQ-REP]**
+
+If the failure detector observes two conflicting headers for height *h*, it should try to verify both. If both are verified it should report evidence.
+
+If the primary reports header *h* and a secondary reports header *h'*,
+     and if *h'* can be verified based on common root of trust, then
+     evidence should be generated; TODO: shall we add "otherwise drop
+     h' and continue normal operation"?
+
+
+## Definitions
+
+- A fixed set of full nodes is provided in the configuration upon
+     initialization. Initially this set is partitioned into
+    -  one full node that is the *primary* (singleton set),
+	-  a set *Secondaries* (of fixed size, e.g., 3),
+	-  a set *FullNodes*.
+- A set *FaultyNodes* of nodes that the light client suspects of being faulty; it is initially empty
+- *State*  is a set of pairs *(fn,h)* where header *h* has been received (and possibly verified with) full node *fn*.
+- The verifier communicates with the primary [[verification]]. Whenever the verifier successfully verifies a header *h* from the primary *p*, it stores
+ *(p,h)*
+ in *State*.
+
+
+## Solution
+
+
+
+### Inter Process Communication
+
+
+
+For the purpose of this light client specification, we assume that the
+     Tendermint Full Node exposes the following functions over
+     Tendermint RPC:
+
+
+```go
+func Commit(addr Address, height int64) (SignedHeader, error)
+```
+
+- Implementation remark
+   - RPC to full node *n* at address *addr*
+- Expected precodnition
+  - header of `height` exists on blockchain
+- Expected postcondition
+  - if *n* is correct: Returns a sound signed header of height `height`
+  from the blockchain if communication is timely (no timeout)
+  - if *n* is faulty: Returns a signed header with arbitrary content
+- Error condition
+   * if *n* is correct: precondition violated or timeout
+   * if *n* is faulty: arbitrary error
+
+#### **[FN-LuckyCase]**:
+The  full node on which the procedure is called remotely is correct and no timeout occurs at the receiver on `Commit`.
+
+#### **[FN-ManifestFaulty]**
+The  full node on which the procedure is called remotely is faulty and a faulty header is received.
+
+----
+
+### Auxiliary Functions (Local)
+
+
+```go
+Add_to_state(addr Address, sh SignedHeader)
+```
+- Expected postcondition
+   - The pair *(addr,sh)* is added to *State*
+
+
+
+```go
+still_punishable(sh SignedHeader) (Boolean)
+```
+- Implementation Remark: it might make sense to check whether the unbonding period is
+  still running although the trusting period is over
+  TODO: fix the period that should be checked. Something between
+  trusting period and unbonding period?
+- Expected postcondition
+    - returns true if misbehavior related to *sh* can still be
+      punished. Can be approximated by *sh.bfttime + unbondingperiod > now*
+
+
+
+```go
+Replace_Secondary(addr Address)
+```
+- Expected precondition
+    - *FullNodes* is nonempty
```

Comment:
Maybe we can state "*FullNodes* is nonempty" as an assumption, for instance, that the address book *FullNodes* is populated by gossip (e.g., peer exchange?).

Reply:
It would be nice. We have to discuss whether this is reasonable. The question is how "light" the client stays if we add this. I guess we said that we keep an addition of peer exchange for later.